impeller/entity/geometry/round_superellipse_geometry.cc

```diff
@@ -0,0 +1,306 @@
+      const Point& point = points[points.size() - i - 1];
+      output.emplace_back(point.y + center.x, point.x + center.y);
+    }
+  }
+}
+
+static Scalar LimitRadius(Scalar corner_radius, const Rect& bounds) {
+  return std::min(corner_radius,
+                  std::min(bounds.GetWidth() / 2, bounds.GetHeight() / 2));
+}
+
+RoundSuperellipseGeometry::RoundSuperellipseGeometry(const Rect& bounds,
+                                                     Scalar corner_radius)
+    : bounds_(bounds), corner_radius_(LimitRadius(corner_radius, bounds)) {}
+
+RoundSuperellipseGeometry::~RoundSuperellipseGeometry() {}
+
+GeometryResult RoundSuperellipseGeometry::GetPositionBuffer(
+    const ContentContext& renderer,
+    const Entity& entity,
+    RenderPass& pass) const {
+  const Size size = bounds_.GetSize();
+  const Point center = bounds_.GetCenter();
+
+  // The full shape is divided into 4 segments: the top and bottom edges come
+  // from two square-like rounded superellipses (width-aligned), while the left
+  // and right squircles come from another two (height-aligned).
+  //
+  // Denote the distance from the center of the square-like squircles to the
+  // origin as `c`. The width-aligned square-like squircle and the
+  // height-aligned one have the same offset in different directions.
+  const Scalar c = (size.width - size.height) / 2;
+
+  // Draw the first quadrant of the shape and store in `points`. It will be
+  // mirrored to other quadrants later.
+  std::vector<Point> points;
+  points.reserve(41);
+
+  DrawOctantSquareLikeSquircle(points, size.width, corner_radius_, Point{0, -c},
+                               false);
+  points.push_back(Point(size / 2) - gap(corner_radius_));  // Point M
+  DrawOctantSquareLikeSquircle(points, size.height, corner_radius_, Point{c, 0},
+                               true);
+
+  static constexpr Point reflection[4] = {{1, 1}, {1, -1}, {-1, -1}, {-1, 1}};
+
+  // Reflect the 1/4 arc into the 4 quadrants and generate the tessellated mesh.
+  // The iteration order is reversed so that the trianges are continuous from
+  // quadrant to quadrant.
+  std::vector<Point> geometry;
+  geometry.reserve(1 + 4 * points.size());
+  geometry.push_back(center);
+  // All arcs include the starting point and exclude the ending point.
+  for (auto i = 0u; i < points.size() - 1; i++) {
+    geometry.push_back(center + (reflection[0] * points[i]));
+  }
+  for (auto i = points.size() - 1; i >= 1; i--) {
+    geometry.push_back(center + (reflection[1] * points[i]));
+  }
+  for (auto i = 0u; i < points.size() - 1; i++) {
+    geometry.push_back(center + (reflection[2] * points[i]));
+  }
+  for (auto i = points.size() - 1; i >= 1; i--) {
+    geometry.push_back(center + (reflection[3] * points[i]));
+  }
+  geometry.push_back(center + points[0]);
+
+  std::vector<uint16_t> indices;
+  indices.reserve(geometry.size() * 3);
+  for (auto i = 2u; i < geometry.size(); i++) {
```

Comment:
Are you generating a triangle fan structure? if so you can use the triangle fan geometry type instead and skip the index buffer (where supported).